[Feature] Add New Question Answering Model

faradawn opened the pull request:
### Description
- Added a question_answering_model.py to trace the model into TorchScript or Onnx format. 
- Added a test file to compare the traced model with the original model's output.

### Issues Resolved
- Issue: https://github.com/opensearch-project/opensearch-py-ml/issues/304
- Old pull request: https://github.com/opensearch-project/opensearch-py-ml/pull/329
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

### Test cases 
All tests passed.
```
test_cases = [
    {
        "question": "Who was Jim Henson?",
        "context": "Jim Henson was a nice puppet"
    },
    {
        "question": "Where do I live?",
        "context": "My name is Sarah and I live in London"
    },
    {
        "question": "What's my name?",
        "context": "My name is Clara and I live in Berkeley."
    },
    {
        "question": "Which name is also used to describe the Amazon rainforest in English?",
        "context": "The Amazon rainforest (Portuguese: Floresta Amazônica or Amazônia; Spanish: Selva Amazónica, Amazonía or usually Amazonia; French: Forêt amazonienne; Dutch: Amazoneregenwoud), also known in English as Amazonia or the Amazon Jungle, is a moist broadleaf forest that covers most of the Amazon basin of South America. This basin encompasses 7,000,000 square kilometres (2,700,000 sq mi), of which 5,500,000 square kilometres (2,100,000 sq mi) are covered by the rainforest. This region includes territory belonging to nine nations. The majority of the forest is contained within Brazil, with 60% of the rainforest, followed by Peru with 13%, Colombia with 10%, and with minor amounts in Venezuela, Ecuador, Bolivia, Guyana, Suriname and French Guiana. States or departments in four nations contain 'Amazonas' in their names. The Amazon represents over half of the planet's remaining rainforests, and comprises the largest and most biodiverse tract of tropical rainforest in the world, with an estimated 390 billion individual trees divided into 16,000 species."
    }
]
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

faradawn committed:
add model and pytest file, passed all tests

Signed-off-by: faradawn <faradawny@gmail.com>

rawwar commented on opensearch_py_ml/ml_models/__init__.py:
```diff
@@ -7,5 +7,6 @@
+from .question_answering_model import QuestionAnsweringModel
 
-__all__ = ["SentenceTransformerModel", "MCorr"]
+__all__ = ["SentenceTransformerModel", "SentenceTransformerModel", "MCorr"]
```
Is this supposed to be `QuestionAnsweringModel`?

rawwar commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,542 @@
+import yaml
+from accelerate import Accelerator, notebook_launcher
+from mdutils.fileutils import MarkDownFile
+# from sentence_transformers import SentenceTransformer
```
Can we remove commented imports?

rawwar commented on tests/ml_models/test_question_answering_pytest.py:
```diff
@@ -0,0 +1,261 @@
+
+default_model_id = "distilbert-base-cased-distilled-squad"
+
+def clean_test_folder(TEST_FOLDER):
```
Can you please take a look at this [link](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) - It talks about how to use temporary files, directories with pytest. Do you think, using these fixtures help us?

rawwar commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,542 @@
+                else self.onnx_zip_file_path
+            )
+
+            # model_zip_file_path = '/Users/faradawn/CS/opensearch-py-ml/opensearch_py_ml/ml_models/question-model-folder/distilbert-base-cased-distilled-squad.zip'
```
Removable comment?

rawwar reviewed (changes requested):
I just reviewed code and hasn't actually tested its workings. Will provide another detailed feedback once i run these locally. Thanks!

faradawn committed:
remove comments and fix init.py

Signed-off-by: faradawn <faradawny@gmail.com>

faradawn committed:
remove model zip path comment

Signed-off-by: faradawn <faradawny@gmail.com>

rawwar commented:
@dhrubo-os , can you please approve tests to run

faradawn committed:
add onnxruntime in requirements-dev.txt

Signed-off-by: faradawn <faradawny@gmail.com>

dhrubo-os commented:
@faradawn I think overall this is a great start. Thanks for raising this PR

1. lint is failing.
2. codecode is already showing some missing coverage tests, please add more unit tests
3. Address PR comments.

We are extending the program few more weeks. So please continue on this PR. Thanks, happy coding.

faradawn commented:
Hi Dhrubo,

Thanks for checking the PR! Will work on it more by fixing the lint and  unit tests!

Thanks,
Faradawn
________________________________
From: Dhrubo Saha ***@***.***>
Sent: Tuesday, December 5, 2023 7:42:26 PM
To: opensearch-project/opensearch-py-ml ***@***.***>
Cc: Faradawn Yang ***@***.***>; Mention ***@***.***>
Subject: Re: [opensearch-project/opensearch-py-ml] [Feature] Add New Question Answering Model (PR #349)


@faradawn<https://urldefense.com/v3/__https://github.com/faradawn__;!!BpyFHLRN4TMTrA!4jzRdXfMXQCAEir1NmEQ_Im18dmrC9o1jFWQtr2hc8BFIG_s59PkaiTbo4Df31yauZY_ghI8VX_HupX1S67BTLRCdiPh$> I think overall this is a great start. Thanks for raising this PR

  1.  lint is failing.
  2.  codecode is already showing some missing coverage tests, please add more unit tests
  3.  Address PR comments.

We are extending the program few more weeks. So please continue on this PR. Thanks, happy coding.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/opensearch-project/opensearch-py-ml/pull/349*issuecomment-1841938846__;Iw!!BpyFHLRN4TMTrA!4jzRdXfMXQCAEir1NmEQ_Im18dmrC9o1jFWQtr2hc8BFIG_s59PkaiTbo4Df31yauZY_ghI8VX_HupX1S67BTDe928QZ$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ARNNCKAGDOSVGEK3QSHA5KDYH7EQFAVCNFSM6AAAAABABQCRZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNBRHEZTQOBUGY__;!!BpyFHLRN4TMTrA!4jzRdXfMXQCAEir1NmEQ_Im18dmrC9o1jFWQtr2hc8BFIG_s59PkaiTbo4Df31yauZY_ghI8VX_HupX1S67BTLDKbXYv$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

mingshl commented:
try running `./gradlew spotlessApply` should help fixing the format,
try running  `./gradlew jacocoTestReport` you can find the test coverage html file in `build/reports/jacoco/test/html/` and you will see what is missing from the unit tests

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+    ) -> str:
+        """
+        Download the model directly from huggingface, convert model to torch script format,
+        zip the model file and its tokenizer.json file to prepare to upload to the Open Search cluster
```
Hi Ming Shi, Thanks for the comments. I am on it!

faradawn committed:
Merge branch 'main' of https://github.com/opensearch-project/opensearch-py-ml into feature/question_answering_model

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+        Download the model directly from huggingface, convert model to torch script format,
+        zip the model file and its tokenizer.json file to prepare to upload to the Open Search cluster
+
+        :param sentences:
```
Changed to optional.

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+            Required, for example  sentences = ['today is sunny']
+        :type sentences: List of string [str]
+        :param model_id:
+            question answering model id to download model from question answerings.
```
Changed to optional.

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+            Required, for example  sentences = ['today is sunny']
+        :type sentences: List of string [str]
+        :param model_id:
+            question answering model id to download model from question answerings.
```
Yes, will download the model with 'model_id' from huggingface.

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+        Download question answering model directly from huggingface, convert model to onnx format,
+        zip the model file and its tokenizer.json file to prepare to upload to the Open Search cluster
+
+        :param model_id:
```
Changed to optional.

faradawn commented on opensearch_py_ml/ml_models/question_answering_model.py:
```diff
@@ -0,0 +1,496 @@
+        zip the model file and its tokenizer.json file to prepare to upload to the Open Search cluster
+
+        :param model_id:
+            question answering model id to download model from question answerings.
```
Yes, will download from huggingface. Thanks!

faradawn committed:
add unit tests and fixed optional parameters

Signed-off-by: faradawn <faradawny@gmail.com>

faradawn commented:
Hi @mingshl,

Thank you for the careful review. I have made changes to the function descriptions, e.g. optional parameters, accordingly.

Hi @dhrubo-os,

I have checked CodeCov's result and added unit tests accordingly. The code is ready for a CodeCov test again.

Thanks,
Faradawn

dhrubo-os commented:
@faradawn let's wrap up the PR? Can you please fix the conflicts?

faradawn committed:
Merge branch 'main' into feature/question_answering_model

Signed-off-by: Faradawn Yang <73060648+faradawn@users.noreply.github.com>

faradawn commented:
Got it, Dhrubo. I will fix the following lint issue. 
```
nox > black --check --target-version=py38 setup.py noxfile.py opensearch_py_ml/ utils/ tests/
would reformat /home/runner/work/opensearch-py-ml/opensearch-py-ml/opensearch_py_ml/ml_models/question_answering_model.py
would reformat /home/runner/work/opensearch-py-ml/opensearch-py-ml/tests/ml_models/test_question_answering_pytest.py
```

dhrubo-os commented:
let's add the changelog file.

faradawn committed:
add to requirements-dev, add to CHANGELOG, fix format with nox

Signed-off-by: faradawn <faradawny@gmail.com>

faradawn commented:
Hi @dhrubo-os, 

Thanks for checking! I have added a missing package in requirement-dev.txt, added to CHANGELOG file, and fixed formating issues.

On my Mac, I only know `pytest` and `nox -rs test`. If there is a more comprehensive test I can run, please let me know!

Thanks.

faradawn committed:
fix sentencetransfoer test of long description

Signed-off-by: faradawn <faradawny@gmail.com>

faradawn commented:
Hi @dhrubo-os, the failing integration test is fixed. There are 4 worklows awaiting approval. If there is anything I can do, please let me know.
